### PR TITLE
Do not return AvailabilitySet if failure domain is set in Machine

### PR DESCRIPTION
**Pull request by enxebre**
It is not a valid config for Azure setting both AvailabilitySet and failure domain, this prevent this from happening for scenarios like externally manage infrastructure where AzureCluster.Status.FailureDomains might not be set but a consumer might still want to set a failure Domain for their pool of Machines

 

 
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/5097

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
If a FailureDomain is set in a Machine, an availability set won't be created for it even if cluster.status.failureDomains is empty
```

### CODE REVIEW

**Comment by enxebre** [comment]
ptal @fxierh @Patryk-Stefanski

File: azure/scope/machine.go [hunk 1]
```diff
@@ -508,7 +508,8 @@ func (m *MachineScope) AvailabilitySetSpec() azure.ResourceSpecGetter {
 func (m *MachineScope) AvailabilitySet() (string, bool) {
 	// AvailabilitySet service is not supported on EdgeZone currently.
 	// AvailabilitySet cannot be used with Spot instances.
-	if !m.AvailabilitySetEnabled() || m.AzureMachine.Spec.SpotVMOptions != nil || m.ExtendedLocation() != nil {
+	if !m.AvailabilitySetEnabled() || m.AzureMachine.Spec.SpotVMOptions != nil || m.ExtendedLocation() != nil ||
```
Comment (jackfrancis) [inline azure/scope/machine.go:511]: I'm curious why these more specific negative conditions aren't all part of the `AvailabilitySetEnabled()` method?

cc @nojnhuh

Comment (nojnhuh) [reply azure/scope/machine.go:511]: I think because `AvailabilitySetEnabled()` is currently defined on the Cluster scope where these other machine-specific options aren't available. I could be convinced that we should move it to the Machine scope though where we do have access to both the Cluster and Machine. I wouldn't block this PR on that refactoring though since I see no other uses of `AvailabilitySetEnabled()` besides here.

Comment (jackfrancis) [reply azure/scope/machine.go:511]: Agree on not blocking, thanks for the pro tip on the scope inheritance.

**Comment by jackfrancis** [comment]
@enxebre thanks for this!

Could you add at least two new test cases to `TestMachineScope_AvailabilitySet` to cover the two new negative OR conditions?

**Comment by enxebre** [comment]
updated @jackfrancis

**Review by willie-yao — APPROVED** [review]
/lgtm
/assign @jackfrancis
